zkevm-circuits/src/bytecode_circuit/circuit.rs

```diff
@@ -0,0 +1,1030 @@
+            );
+
+            cb.require_equal(
+                "cur.value == cur.length",
+                meta.query_advice(bytecode_table.value, Rotation::cur()),
+                meta.query_advice(length, Rotation::cur()),
+            );
+
+            cb.gate(and::expr(vec![
+                meta.query_fixed(q_enable, Rotation::cur()),
+                not::expr(meta.query_fixed(q_last, Rotation::cur())),
+                is_header(meta),
+            ]))
+        });
+
+        // When is_byte ->
+        // assert push_data_size_table_lookup(cur.value, cur.push_data_size)
+        // assert cur.is_code == (cur.push_data_left == 0)
+        meta.create_gate("Byte row", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            cb.require_equal(
+                "cur.is_code == (cur.push_data_left == 0)",
+                meta.query_advice(bytecode_table.is_code, Rotation::cur()),
+                push_data_left_is_zero.clone().is_zero_expression,
+            );
+
+            cb.gate(and::expr(vec![
+                meta.query_fixed(q_enable, Rotation::cur()),
+                not::expr(meta.query_fixed(q_last, Rotation::cur())),
+                is_byte(meta),
+            ]))
+        });
+        meta.lookup_any(
+            "push_data_size_table_lookup(cur.value, cur.push_data_size)",
+            |meta| {
+                let enable = and::expr(vec![
+                    meta.query_fixed(q_enable, Rotation::cur()),
+                    not::expr(meta.query_fixed(q_last, Rotation::cur())),
+                    is_byte(meta),
+                ]);
+
+                let lookup_columns = vec![value, push_data_size];
+
+                let mut constraints = vec![];
+
+                for i in 0..PUSH_TABLE_WIDTH {
+                    constraints.push((
+                        enable.clone() * meta.query_advice(lookup_columns[i], Rotation::cur()),
+                        meta.query_fixed(push_table[i], Rotation::cur()),
+                    ))
+                }
+                constraints
+            },
+        );
+
+        // When is_header_to_header or q_last ->
+        // assert cur.length == 0
+        // assert cur.hash == EMPTY_HASH
+        meta.create_gate("Header to header row", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            cb.require_zero(
+                "cur.length == 0",
+                meta.query_advice(length, Rotation::cur()),
+            );
+
+            let empty_hash = RandomLinearCombination::<F, 32>::random_linear_combine_expr(
+                EMPTY_HASH_LE.map(|v| Expression::Constant(F::from(v as u64))),
+                &challenges.evm_word_powers_of_randomness::<32>(),
+            );
+
+            cb.require_equal(
+                "assert cur.hash == EMPTY_HASH",
+                meta.query_advice(bytecode_table.code_hash, Rotation::cur()),
+                empty_hash,
```

Comment:
I remember we have `cb.empty_hash_rlc()`, but that method was available for `ConstraintBuilder` but not for `BaseConstraintBuilder`.
No change is needed here.